dill/_utils.py

```diff
@@ -0,0 +1,207 @@
+    def __getattr__(self, key):
+        # This is called only if dict.__getattribute__(key) fails.
+        try:
+            return self[key]
+        except KeyError:
+            raise AttributeError("'AttrDict' object has no attribute %r" % key)
+    def __setattr__(self, key, value):
+        self._check_attr(key)
+        self[key] = value
+    def __delattr__(self, key):
+        self._check_attr(key)
+        del self[key]
+    def __reduce__(self):
+        return type(self), (dict(self),)
+
+
+### Namespace filtering
+import re
+from dataclasses import InitVar, dataclass, field, fields
+from collections import abc, namedtuple
+from enum import Enum
+from functools import partialmethod
+from itertools import filterfalse
+from typing import Callable, Iterable, Pattern, Set, Tuple, Union
+
+RuleType = Enum('RuleType', 'EXCLUDE INCLUDE', module=__name__)
+NamedObj = namedtuple('NamedObj', 'name value', module=__name__)
+
+Filter = Union[str, Pattern[str], int, type, Callable]
+Rule = Tuple[RuleType, Union[Filter, Iterable[Filter]]]
+
+def isiterable(arg):
+    return isinstance(arg, abc.Iterable) and not isinstance(arg, (str, bytes))
+
+@dataclass
+class ExcludeFilters:
+    ids: Set[int] = field(default_factory=set)
+    names: Set[str] = field(default_factory=set)
+    regex: Set[Pattern[str]] = field(default_factory=set)
+    types: Set[type] = field(default_factory=set)
+    funcs: Set[Callable] = field(default_factory=set)
+
+    @property
+    def filter_sets(self):
+        return tuple(field.name for field in fields(self))
+    def __bool__(self):
+        return any(getattr(self, filter_set) for filter_set in self.filter_sets)
+    def _check(self, filter):
+        if isinstance(filter, str):
+            if filter.isidentifier():
+                field = 'names'
+            else:
+                filter, field = re.compile(filter), 'regex'
+        elif isinstance(filter, re.Pattern):
+            field = 'regex'
+        elif isinstance(filter, int):
+            field = 'ids'
+        elif isinstance(filter, type):
+            field = 'types'
+        elif callable(filter):
+            field = 'funcs'
+        else:
+            raise ValueError("invalid filter: %r" % filter)
+        return filter, getattr(self, field)
+    def add(self, filter):
+        filter, filter_set = self._check(filter)
+        filter_set.add(filter)
+    def discard(self, filter):
+        filter, filter_set = self._check(filter)
+        filter_set.discard(filter)
+    def remove(self, filter):
+        filter, filter_set = self._check(filter)
+        filter_set.remove(filter)
+    def update(self, filters):
+        for filter in filters:
+            self.add(filter)
+    def clear(self):
+        for filter_set in self.filter_sets:
+            getattr(self, filter_set).clear()
+    def add_type(self, type_name):
+        import types
+        name_suffix = type_name + 'Type' if not type_name.endswith('Type') else type_name
```

Comment:
Most types are defined in `types` *or* `builtins`, but there's also types defined in other modules, and types that aren't explicitly named.  Is your little bit of convenience here a good idea?

Reply:
Changed to use `dill._dill._reverse_typemap`. I think it's useful, and doesn't hurt anyone. The usage is simple:
```python
>>> import dill
>>> filters = dill.settings['dump_session']
>>> filters.exclude.add_type('function')  # 'Function' and 'FunctionType' also work
```

Reply:
After thinking more about it, I decided to replace the `FilterSet.add_type()` method by a special string syntax for the `FilterSet.add()` method. It didn't hurt the filter type logic and it's better to have a single entry point. Now the usage is `filters.add('type:function')` instead of `filters.add_type('function')`.